# Heart Disease K-Means Clustering Analysis

## Project Overview
This notebook performs unsupervised machine learning analysis using K-Means clustering on a heart disease dataset. The goal is to identify natural groupings of patients based on their clinical features.

**Dataset**: Heart Disease Dataset (920 patient records)

**Kaggle Link**: https://www.kaggle.com/competitions/k-means-clustering-for-heart-disease-analysis/overview

**Objective**: Use K-Means clustering to group patients into distinct clusters based on cardiovascular health indicators.

---

## Phase 1: Preprocessing & EDA

### Step 1: Setup & Data Loading

In [1]:
# Import necessary libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import warnings

# Machine Learning libraries
from sklearn.preprocessing import StandardScaler, LabelEncoder
from sklearn.cluster import KMeans
from sklearn.decomposition import PCA
from sklearn.metrics import silhouette_score, silhouette_samples

# Configure visualization settings
plt.style.use('seaborn-v0_8-darkgrid')
sns.set_palette('husl')
warnings.filterwarnings('ignore')

# Set display options
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', 100)
pd.set_option('display.float_format', lambda x: '%.3f' % x)

print("✓ Libraries imported successfully")

✓ Libraries imported successfully


In [2]:
# Load the dataset
df = pd.read_csv('heart_disease.csv')

# Display first few rows
print("Dataset loaded successfully!\n")
print(f"Dataset shape: {df.shape[0]} rows × {df.shape[1]} columns\n")
df.head(10)

Dataset loaded successfully!

Dataset shape: 920 rows × 15 columns



id  age     sex    dataset               cp  trestbps    chol    fbs  \
0   0   63    Male  Cleveland   typical angina   145.000 233.000   True   
1   1   67    Male  Cleveland     asymptomatic   160.000 286.000  False   
2   2   67    Male  Cleveland     asymptomatic   120.000 229.000  False   
3   3   37    Male  Cleveland      non-anginal   130.000 250.000  False   
4   4   41  Female  Cleveland  atypical angina   130.000 204.000  False   
5   5   56    Male  Cleveland  atypical angina   120.000 236.000  False   
6   6   62  Female  Cleveland     asymptomatic   140.000 268.000  False   
7   7   57  Female  Cleveland     asymptomatic   120.000 354.000  False   
8   8   63    Male  Cleveland     asymptomatic   130.000 254.000  False   
9   9   53    Male  Cleveland     asymptomatic   140.000 203.000   True   

          restecg  thalch  exang  oldpeak        slope    ca  \
0  lv hypertrophy 150.000  False    2.300  downsloping 0.000   
1  lv hypertrophy 108.000   True    1.500         flat 3.000   
2  lv hypertrophy 129.000   True    2.600         flat 2.000   
3          normal 187.000  False    3.500  downsloping 0.000   
4  lv hypertrophy 172.000  False    1.400    upsloping 0.000   
5          normal 178.000  False    0.800    upsloping 0.000   
6  lv hypertrophy 160.000  False    3.600  downsloping 2.000   
7          normal 163.000   True    0.600    upsloping 0.000   
8  lv hypertrophy 147.000  False    1.400         flat 1.000   
9  lv hypertrophy 155.000   True    3.100  downsloping 0.000   

                thal  
0       fixed defect  
1             normal  
2  reversable defect  
3             normal  
4             normal  
5             normal  
6             normal  
7             normal  
8  reversable defect  
9  reversable defect

**Key observations:**
- Dataset contains **920 patient records** with **15 columns**
- Features include both numerical (age, blood pressure, cholesterol) and categorical (sex, chest pain type, ECG results) variables
- The dataset appears to be from multiple sources (Cleveland, VA Long Beach datasets)

**Feature descriptions:**
- `id`: Patient identifier
- `age`: Age in years
- `sex`: Male/Female
- `dataset`: Source dataset
- `cp`: Chest pain type (typical angina, atypical angina, non-anginal, asymptomatic)
- `trestbps`: Resting blood pressure (mm Hg)
- `chol`: Serum cholesterol (mg/dl)
- `fbs`: Fasting blood sugar > 120 mg/dl (True/False)
- `restecg`: Resting electrocardiographic results
- `thalch`: Maximum heart rate achieved
- `exang`: Exercise induced angina (True/False)
- `oldpeak`: ST depression induced by exercise
- `slope`: Slope of peak exercise ST segment
- `ca`: Number of major vessels colored by fluoroscopy
- `thal`: Thalassemia type